src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc

```diff
@@ -1163,36 +1163,56 @@ static int glb_pick_locked(grpc_exec_ctx *exec_ctx, grpc_lb_policy *pol,
                            "won't work without it. Failing"));
     return 0;
   }
-
   glb_lb_policy *glb_policy = (glb_lb_policy *)pol;
-  bool pick_done;
-
+  bool pick_done = false;
   if (glb_policy->rr_policy != NULL) {
-    if (GRPC_TRACER_ON(grpc_lb_glb_trace)) {
-      gpr_log(GPR_INFO, "grpclb %p about to PICK from RR %p",
-              (void *)glb_policy, (void *)glb_policy->rr_policy);
-    }
-    GRPC_LB_POLICY_REF(glb_policy->rr_policy, "glb_pick");
+    const grpc_connectivity_state rr_connectivity_state =
+        grpc_lb_policy_check_connectivity_locked(exec_ctx,
+                                                 glb_policy->rr_policy, NULL);
+    // The glb_policy->rr_policy may have transition to SHUTDOWN but the
+    // callback registered to capture this event
+    // (glb_rr_connectivity_changed_locked) may not have been invoked yet. We
+    // need to make sure we aren't trying to pick from a RR policy instance
+    // that's in shutdown.
+    if (rr_connectivity_state == GRPC_CHANNEL_SHUTDOWN) {
+      if (GRPC_TRACER_ON(grpc_lb_glb_trace)) {
+        gpr_log(GPR_INFO,
+                "grpclb %p NOT picking from from RR %p: RR conn state=%s",
+                (void *)glb_policy, (void *)glb_policy->rr_policy,
+                grpc_connectivity_state_name(rr_connectivity_state));
+      }
+      // Attempt to switch over to a pending update.
```

Comment:
In what situation will this be useful?  If we actually have a pending update, won't we have already applied it?  Moreover, if there is a pending update, won't we switch to it as soon as the connectivity callback is invoked anyway?

Reply:
The intent is to take action ASAP. While it's true that an eventual  `glb_rr_connectivity_changed_locked` would set `rr_policy` to `NULL`, it'd then take _another_ pick attempt for the `else {  // glb_policy->rr_policy == NULL` branch in `glb_pick_locked` (line 1213) to add the pick to the pending list. Moreover, a new RR policy won't be created until a new serverlist is received (the `glb_rr_connectivity_changed_locked` function does not perform any handover or processing of pending updates when the RR's connectivity state is `SHUTDOWN`).

Reply:
My concern isn't about adding the pick to the pending pick list; that's clearly the right thing to do here.  My concern is about calling `rr_handover_locked()`.  I could be missing something here, but I think this is both pointless and possibly dangerous.

The comment here implies that the reason for calling `rr_handover_locked()` is to switch to a pending update (i.e., new list of backend addresses for the RR policy).  However, it's not clear to me when that there would ever be an update available when we get here.  When we get a new serverlist from the balancer, we call `rr_handover_locked()` immediately (see `lb_on_response_received_locked()`), so there's basically no way this can run between when we get a new serverlist and when `rr_handover_locked()` is called on that serverlist.  If this runs before we get a new serverlist, then all we're doing is recreating the RR policy with the same set of addresses that went bad; if it runs afterwards, it's a no-op, because it's updating the RR policy with the list of addresses it already has.

Furthermore, it's not clear to me that it's actually safe to give an update to the RR policy when it's in state SHUTDOWN.  I don't think we've tested that case, and it seems to me that it's conceptually not something we should ever do.  In fact, I would argue that we should have an assertion in `rr_update_locked()` that the policy is not being shut down.  (Same for all LB policies, actually.)

And finally, even if it is safe to call update on a policy that's shutting down, it seems completely pointless to do so, because we'll still wind up unreffing the RR policy very soon anyway, as soon as the connectivity state callback comes back telling us that the RR policy is in state SHUTDOWN.  It's not clear to me whether the RR policy will make any forward progress in that case, but this is yet another code-path that seems untested and dangerous to me.

Wouldn't it make sense to just put the pick on the pending list and wait for the the normal mechanism to recreate the RR policy and drain the pending pick list?  In other words, we treat the case where the RR policy is in state SHUTDOWN the same as if it was NULL.  That seems like a much safer approach here.

Reply:
I see your point, agreed. Removed.